Sources/Deferred/ReadWriteLock.swift

```diff
@@ -0,0 +1,258 @@
+    }
+
+    /// Call `body` with a writing lock.
+    ///
+    /// The given function is guaranteed to be called exclusively.
+    ///
+    /// - parameter body: A function that writes a value while locked, then returns some value.
+    /// - returns: The value returned from the given function.
+    public func withWriteLock<Return>(@noescape body: () throws -> Return) rethrows -> Return {
+        // spin until we acquire write lock
+        repeat {
+            let state = __bnr_atomic_load_32(_state)
+
+            // if there are no readers and no one holds the write lock, try to grab the write lock immediately
+            if (state == 0 || state == Masks.WRITER_WAITING_BIT) &&
+                __bnr_atomic_compare_and_swap_32(_state, state, Masks.WRITER_BIT) {
+                    break
+            }
+
+            // If we get here, someone is reading or writing. Set the WRITER_WAITING_BIT if
+            // it isn't already to block any new readers, then wait a bit before
+            // trying again. Ignore CAS failure - we'll just try again next iteration
+            __bnr_atomic_or_32(_state, Masks.WRITER_WAITING_BIT)
+        } while true
+
+        defer {
+            // unlock by clearing everything except (possibly) WRITER_WAITING_BIT,
+            // which will only be set if another writer is already here and
+            // waiting (which will keep out readers)
+            __bnr_atomic_and_32(_state, Masks.WRITER_WAITING_BIT)
+        }
+
+        // write lock acquired - run block
+        return try body()
+    }
+
+    /// Call `body` with a reading lock.
+    ///
+    /// The given function may be called concurrently with reads on other threads.
+    ///
+    /// - parameter body: A function that reads a value while locked.
+    /// - returns: The value returned from the given function.
+    public func withReadLock<Return>(@noescape body: () throws -> Return) rethrows -> Return {
+        // spin until we acquire read lock
+        repeat {
+            let oldCount = __bnr_atomic_load_32(_state) & Masks.MASK_READER_BITS
+            let newCount = oldCount.successor()
+
+            // if there is no writer and no writer waiting, try to increment reader count
+            if __bnr_atomic_compare_and_swap_32(_state, oldCount, newCount) {
+                break
+            }
+        } while true
+        
+        defer {
+            // decrement reader count
+            repeat {
+                let state = __bnr_atomic_load_32(_state)
+                let readerCount = state & Masks.MASK_READER_BITS
+                
+                // sanity check that we have a positive reader count before decrementing it
+                assert(readerCount > 0, "unlocking read lock - invalid reader count")
+                
+                // desired new state: 1 fewer reader, preserving whether or not there is a writer waiting
+                let newState = (readerCount - 1) | (state & Masks.WRITER_WAITING_BIT)
```

Comment:
Seems GitHub has misidentified the file move + changes as a new file. :disappointed: Odd, because Git doesn't for the changes. Anyway, [the original implementation](https://github.com/bignerdranch/Deferred/pull/61/files#diff-1311f678fc962b574d7935db66cb98faL133).